src/nerve_tools/envoy.py

```diff
@@ -0,0 +1,137 @@
+import copy
+import logging
+import re
+from typing import cast
+from typing import Mapping
+from typing import Iterable
+from typing import Optional
+from typing import Dict
+
+import requests
+
+from nerve_tools.config import CheckDict
+from nerve_tools.config import ListenerConfig
+from nerve_tools.config import ServiceInfo
+from nerve_tools.config import SubSubConfiguration
+from nerve_tools.util import get_host_ip
+
+
+INGRESS_LISTENER_REGEX = re.compile(r'^\S+\.\S+\.(?P<port>\d+)\.ingress_listener$')
+
+
+def _get_envoy_listeners_from_admin(admin_port: int) -> Mapping[str, Iterable[ListenerConfig]]:
+    try:
+        return requests.get(f'http://localhost:{admin_port}/listeners?format=json').json()
+    except Exception as e:
+        logging.warning(f'Unable to get envoy listeners: {e}')
+        return {}
+
+
+def get_envoy_ingress_listeners(admin_port: int) -> Mapping[int, int]:
+    """Compile a mapping of "service's local listening port" -> "the corresponding Envoy
+    ingress port".
+
+    This will be used to determine the Envoy ingress port for a given service's actual port.
+    """
+    envoy_listeners: Dict[int, int] = {}
+    envoy_listeners_config = _get_envoy_listeners_from_admin(admin_port)
+
+    for listener in envoy_listeners_config.get('listener_statuses', []):
+        result = INGRESS_LISTENER_REGEX.match(listener['name'])
+        if result:
+            local_host_port = result.group('port')
+            try:
+                envoy_listeners[int(local_host_port)] = \
+                    int(listener['local_address']['socket_address']['port_value'])
+            except KeyError:
+                # If there is no socket_address and port_value, skip this listener
+                pass
+    return envoy_listeners
+
+
+def _get_envoy_service_info(
```

Comment:
since we're calling this in configure_nerve.py, let's get rid of the underscore

Reply:
Fixed